src/settings/target/mod.rs

```diff
@@ -266,88 +247,73 @@ impl Manifest {
+            (None, None) => (false, None),
+        };
 
-            // environment configuration
-            Some(environment) => {
-                if let Some(wdd) = environment.workers_dev {
-                    if wdd && environment.route.is_some() {
-                        failure::bail!(wdd_failure)
-                    }
-                    wdd
-                } else if let Some(wdd) = self.workers_dev {
-                    if wdd && environment.route.is_some() {
-                        false // allow route to override workers_dev = true if wdd is inherited
+        // override top level with environment
+        let (workers_dev, route) = if let Some(env) = &environment {
+            log::debug!("env workers_dev: {:?}", env.workers_dev);
+            log::debug!("env route: {:?}", env.route);
+            match (env.workers_dev, env.route.clone()) {
+                (None, Some(route)) => {
+                    if top_workers_dev && route.is_empty() {
+                        failure::bail!(empty_route_failure)
                     } else {
-                        wdd // inherit from top level
+                        (false, Some(route))
                     }
-                } else {
-                    false // if absent -> false
                 }
-            }
-        };
-
-        let route = if let Some(environment) = environment {
-            if let Some(route) = &environment.route {
-                if let Some(wdd) = environment.workers_dev {
-                    if wdd {
-                        failure::bail!(wdd_failure);
+                (Some(workers_dev), None) => (workers_dev, None),
+                (Some(workers_dev), Some(route)) => {
+                    if route.is_empty() && workers_dev {
+                        failure::bail!(empty_route_failure)
                     }
+                    (workers_dev, Some(route))
                 }
-                Some(route.clone())
-            } else {
-                None
+                (None, None) => (top_workers_dev, top_route),
             }
         } else {
-            self.route.clone()
+            (top_workers_dev, top_route)
         };
 
-        Ok((route, workers_dev))
+        log::debug!("negotiated workers_dev: {}", workers_dev);
+        log::debug!("negotiated route: {:?}", route);
+
+        match (workers_dev, route) {
+            (true, None) => Ok(None),
+            (true, Some(route)) => {
+                if route.is_empty() {
+                    Ok(None)
+                } else {
+                    failure::bail!(conflicting_targets_failure)
+                }
+            }
+            (false, Some(route)) => {
+                if route.is_empty() {
+                    failure::bail!(pick_target_failure)
+                } else {
+                    Ok(Some(route))
+                }
+            }
+            (false, None) => failure::bail!(pick_target_failure),
+        }
```

Comment:
I feel like this match statement can be completely consolidated with the one above :) 

But beyond this, I think we could do away with matches entirely and just rely on `is_none()` and `is_some()` to deduce which outcome we have for deployment target. I think this would be ultimately more readable than very long match statements and very explicitly display the conditions leading to a truth table assessment (they would all be in the conditional clause, which could really help with readability).

Reply:
I don't think that `is_none()` and `is_some()` is what we want here because they always need to be paired with `unwrap()`, not going to be any more readable imo